### ISSUE TITLE
Compiling

### ISSUE THREAD

**Issue by BeaconDev** [OpenXRay/xray-16#3]
How do we go about creating a working executable or set of executables from the source? I'd be intrigued to try out the supposed new grass shadows, and then start looking at making changes of my own.

**Comment by nitrocaster** [comment]
First of all, I would like to use my CS repository style here (docs, version plan, coding conventions).
Then apply changes I've already done (+ your changes, if there are ones).
Note that we should always have good master branch (though my repo wasn't actually good all the time). 'Good' means that:
- it can be compiled without errors
- it runs without critical errors (at least doesn't add new ones)
- it is compliant with the coding conventions we use

To all contributors: Please, fill in your email addresses so we can contact each other personally.

**Comment by BeaconDev** [comment]
Fine by me, go ahead and set that up if you like. If it looks good and works efficiently I don't see why we can't use your style.

Completely agree with a stable master branch.

You can contact me on CoPRedux@googlemail.com, or probably better, if you have Steam, at http://steamcommunity.com/id/beacon

**Comment by ghost** [comment]
In terms of compiling, I have not tried compiling yet, I just wanted to port over my changes from CS first which I know will work fine with COP.

I completely agree. I think we need to agree on which COP repository to use first.
There are two that come to mind:
1) The official 1.6.02 leak with fixes for VS2008 (that's the one uploaded here)
2) https://bitbucket.org/redpython/xray-cop/branch/1.6.02 

I also think we really should see if we can use VS2013 instead of 2008. Nitrocaster, I know you managed to do that with CS, and I have no idea how to do it at all. Would that be something you'd be interested in doing? You could pick what you want as the main branch then update it to 2013, then we can all work together on changes that follow the same style you use.

My email is matthew.w.swartz@gmail.com

I think what I'm going to do until we get all that done is port over all the changes I did for CS (such as the grass shadows by K.D.), into a separate branch and have only the files changed in there, which would allow for easy merging for features people want or don't want.

**Comment by BeaconDev** [comment]
What are the differences on the two repos you listed?

**Comment by nitrocaster** [comment]
For emails: I meant you [profile settings](https://github.com/settings/profile)

For compiling: It is bad idea to commit code that you didn't try to compile. Things are not supposed to work this way, we need 'good' master branch. When I found CS sources, I spent about two days to make sure it can be compiled.

For reference code: I suggest we use original sources for VS2008. Reason: no new bugs, only old well known ones.

For using VS2013: There's no troubles to use VS2013 - just convert solution and projest files. There may be some problems with virtual directories in projects - that have to be fixed manually (awful routine, but it happens).

**Comment by ghost** [comment]
Works for me. Just filled in my email.
I agree, I shouldn't have suggested on making changes when I have yet to even compile it.

I may not have enough time today (have to work in a few hours), but I'll try to convert the project to VS2013 today then compile it. If it works I'll update things and post instructions the same way you did at your CS repo.

Also, for some reason I had two upload the code in two parts using the Github utility as it would crash after around 7,500 files. Is there an alternative way of committing to Github?

**Comment by nitrocaster** [comment]
So, I'm going to dig into these sources, and once I get them all compiled I'll push them.
Meanwhile I suggest you to learn what's new in xray 1.6 comparing to 1.5 and schedule your tasks in the version plan.

**Comment by nitrocaster** [comment]
@Swartz27, I understood from your previous post that you ask is I would get COP sources compiled under VS2013, so I'm ok with that. Are you going to do it by yourself?

**Comment by ghost** [comment]
I was going to but if you're willing to do it I'll let you, then you can push it, and we can use what you've done as a base.

I realized I have very little time today, I'll probably jump back on in roughly 7 hours and see if anyone has had any luck. I'll go over the proposed features list again and tidy it up after I've compared COP to CS source and then put them in order.

**Comment by nitrocaster** [comment]
> Also, for some reason I had two upload the code in two parts using the Github utility as it would crash after around 7,500 files. Is there an alternative way of committing to Github?

Actually, I don't use Github utility for committing. I prefer git add + git commit + git push. If you're interested I use [WinCmp (Compare it!)](http://www.grigsoft.com/wincmp3.htm) as diff/merge tool.
I have [simple batch file](https://dl.dropboxusercontent.com/u/63973281/dev/xray/tools/gdiff.cmd) that allows me to do these things:
- gdiff <filenames>: run difftool on unstaged changes
- gdiff -c <filenames> : run difftool on staged changes

**Comment by LoNer1** [comment]
Good to be hearing ablout the structure and how things are going to unpack from here on.

For my email, its "loner_dude@hotmail.com" for anything modding related. Looking forward to our cooperation, gents!

Date: Thu, 18 Sep 2014 09:07:49 -0700

**Comment by nitrocaster** [comment]
To @Swartz27 : What are 'cs' and 'patch-1' branches for?

**Comment by ghost** [comment]
I don't remember creating either branch, though it looks like someone removed "patch-1". The CS branch could have been where I had forked your CS repo but it can't be because there's no CS files there. I'll remove it.

**Comment by ghost** [comment]
Bangalore, we're waiting on Nitrocaster to update the source here to Visual Studio 2013. I recommend using that once it's out rather than messing with these other revisions.

**Comment by ghost** [comment]
I suggest to the feature list to implement luajit 2.0.3, for the replace of luajit 1.1.4.:
http://luajit.org/download.html

Luajit above version 2.0.0 compiles lua scripts much faster, it could remove microstutter, you can compare here the performance difference:
http://luajit.org/performance_x86.html

**Comment by BeaconDev** [comment]
100% yes

also, nice to see you here Bangalore!

**Comment by ghost** [comment]
I agree completely, I simply hadn't added the luajit version number in the proposed engine changes, but I had meant the version you speak of.

**Comment by nitrocaster** [comment]
To @Bangalore1010  : r8384 is kinda experimental version.

**Comment by ghost** [comment]
@Bangalore1010 : Did you use the vs2008 version from the Github here or did you download it from the post on gameru? Just wondering because I encountered a bug trying to compile the source here, but I'm pretty sure it's the same version. Either way, downloaded it again from gameru. I think the reason I couldn't get it to compile was I don't have SP1 installed for VS2008 :P

**Comment by tsf00181** [comment]
Do you have a link for the CS source code?

Google translate is playing hell with me on the gameru.net site so I'm
having trouble finding it.

**Comment by ghost** [comment]
@tsf00181 : http://www.gameru.net/forum/index.php?s=297823ebaced19091f6c36b012189218&showtopic=55777&st=1700&p=1489128&#entry1489128

@Bangalore1010 : Did you encounter an error when compiling zlib that mentioned "_vsnprintf" being redefined? Just wondering, because I fixed it by going into zutil.h and commenting out lines 197 to 202.
I'm compiling xrgame (again) because I forgot to include the DirectX includes and libraries with it :p I'm expecting this will work fine though as it's the last file to compile and they've all compiled fine so far.

I'll post an update later as I have to go.

**Comment by andrew-boyarshin** [comment]
I'd like to work on LuaJIT and LuaBind, when CoP sources will become compilable.

**Comment by nitrocaster** [comment]
I'd like to suggest to rename this repo to "xray-16", if you don't mind.

**Comment by ghost** [comment]
I don't mind at all. In fact, nitrocaster, is it possible to transfer ownership of a repo? You have a good system with certain coding conventions, you're the top programmer here, etc. I think you should be in charge with us contributing.

Also, unless you've gotten any farther on compiling it in VS2013, I think maybe we're good with leaving it at VS2008. I downloaded and compiled what Bangalore posted to be extra careful, beside the one modification I had to make which I mentioned, it worked fine (well, the R4 renderer isn't working at the moment. It's telling me my video card doesn't support it. So I need to figure out what's causing that which is probably me using a newer DirectX sdk than the one we're supposed to use).

I'm actually going to delete the source code currently up and re-upload with the structure Bangalore is using once I can verify R4 works correctly. It's cleaner to post it that way.
Then I think I'm going to grab the coding convention and commit convention documents you have and post them here as well if you don't mind.

@STALKER2010 : Please do work on those, that would be great. With LUAJit are you going to be using the newest version, as that is what we would like to use.It means re-doing the vanilla LUA scripts too in order to work with the new version though.

**Comment by nitrocaster** [comment]
> is it possible to transfer ownership of a repo?

AFAIK, it's not necessary since I'm a contributor. Anyway, you can do it via support.

For migrating to VS2013 - work in progress, just wait a little.

For R4 renderer: do you have DX11-compatible GPU? If so, that is a problem in shader that I've already fixed (in my xray 1.5 repo).

For this:

> I'm actually going to delete the source code currently up and re-upload with the structure Bangalore is using

You uploaded the same code Bangalore use (vs2008.fixed). The only difference is that you missed SDK stuff. I'll take care of it, keep calm and don't delete anything :)

**Comment by nitrocaster** [comment]
Wow, it's not that I thought. Try to investigate.

**Comment by ghost** [comment]
My guess is that the SDK isn't up to date for DX11. I'm going to try a later revision of the DirectX SDK and see if that solves it.
If that doesn't help I'll compile everything under debug.

_Edit_
The SDK is the last one before it becomes Windows SDK for Win8 which the GSC people obviously didn't use. Time to debug.

**Comment by ghost** [comment]
Looks like it doesn't want to be debugged: http://pastebin.com/Pr83gBWe

Just noticed something else. Every time it crashes with that error it creates the precompiled shader for the r4 dumb.ps in your gamedata folder. WTF is that all about?

**Comment by nitrocaster** [comment]
> The xrGame.dll is a hell a lot of time to compile.

That's why we should split it to multiple modules (maybe DLL + static libraries).

**Comment by nitrocaster** [comment]
I've just noticed _GPA_ENABLED define. Maybe it's redundant (since VS2013 has really good DirectX debugging tools). Anyway, I gonna move it from source code to project additional defines.

**Comment by nitrocaster** [comment]
> Just noticed something else. Every time it crashes with that error it creates the precompiled shader for the r4 dumb.ps in your gamedata folder. WTF is that all about?

COP caches compiled shaders in $app_data_root$/shaders_cache. gamedata directory should be read-only.

**Comment by nitrocaster** [comment]
This have to be debugged (see d3dcompiler output).

**Comment by nitrocaster** [comment]
This is bad solution - you don't know what causes that problem so you wont be able to fix it when it breaks again. We'll need to get this working with new DX SDK.

I suppose this problem is quite simple to debug, try to fix it by yourself.

**Comment by ghost** [comment]
I can't fix it, i'm not a programmer, sorry. I'm just a beginner with enquiry, so maybe it would be better to remove me from the collaborators.

**Comment by ghost** [comment]
@Bangalore1010 : You're wanted here, you may not be a programmer but I added you as a collaborator so you could make engine suggestions. It looks like Alundaio might be back too so maybe you can convince him to do the same? I don't think he wants to do any programming, but maybe he'd do that.

As for the shader problem and the latest DXSDK, it is a problem as when DirectX 12 comes out it would be cool to add it to the source, and we won't be able to if we can't use the newest SDK that now comes with the Windows SDK. I have an idea on what the problem might be (has to do with the shaders, not the code) so let me try it and I'll report back.

**Comment by nitrocaster** [comment]
Little update. I'm trying to remove all original 3rd party code from the repository (from SDK dir) to reduce its size.
Already done with:
- libogg
- libtheora
- libvorbis
- FreeMagic

**Comment by ghost** [comment]
That's great news :)

I'm going to try and reorganize things on here tonight and tomorrow depending on available time. 

I wasn't able to solve the problem through the shaders like I thought I could. The big problem is that they use precompiled shaders for everything, so there's no knowing what settings were used when they compiled them (for instance the DX11 renderer uses R1 lights for campfires for some reason unknown). I did compare files between the two DirectX sdk's (Aug 2009 vs Jun 2010) and noticed that in DX11 a lot of data was moved from a few files to a few others, though it remains the same data. It probably is a code thing and I'll try and debug it tomorrow.

In the meantime, Bangalore was right that Aug 2009 will compile R4 just fine, but obviously that's a temporary fix if we want to implement DX12 way in the future, but I also don't consider it the end of the world if we don't get DX12. 

I made two small changes to a seperate direcory I have just for screwing around with the game code. It's a good thing I didn't commit those grass shadows and actor shadows, as the format (for actor shadows anyway) is now different. I fixed it and both features look great in R2-R4, but it's a good thing I tested them.

One thing I think is really important to do is get rid of these goddamn pre-compiled shaders, they're a pain in my ass. I asked team Dezowave what they thought of using them, and they said while developing Lost Alpha they tried adopting COP's precompiled shader system, but when they tested it vs. non-compiled shaders they noticed very little difference in performance, so we wouldn't be losing anything and R4 might look a bit better.

**Comment by nitrocaster** [comment]
> One thing I think is really important to do is get rid of these goddamn pre-compiled shaders, they're a pain in my ass.

Precompiled shaders save loading time. What exact problem do you have?

**Comment by ghost** [comment]
It may save loading time, but it makes it difficult for people to add new shaders as if the precompiled shaders have data that conflicts it will override it (and since we don't know what settings they used for their precompiled shaders, we can't compile our own. A good example of this is objects/r3/accum_sun_near.ps/ . It has a variety of files in there, all with different file sizes).

On top of that, R4 has had a long standing issue where it uses R1 lighting around things like campfires. I imagine that would be in the precompiled shaders, not the source code.

You're welcome to keep precompiled shaders, but I think I'll remove that feature on my end (and won't commit it here). It's just an experiment to see if it helps with R4 for instance.

**Comment by BeaconDev** [comment]
I'd definitely chime in and put in some support for non-precompiled shaders, the campfire issue is a real pain in my ass.

**Comment by nitrocaster** [comment]
This should be investigated and fixed, not thrown away. We'll face lots of issues that have to be fixed too (but cannot be thrown away like this).

**Comment by andrew-boyarshin** [comment]
@nitrocaster when will CoP sources will be adapted for VS2013? I use it, and I don't want to install VS2008. Please, make it compilable under VS2013 and without virtual drives magic. Thanks.

**Comment by nitrocaster** [comment]
WIP. Currently I'm fixing project dependencies and include/library paths. Final result is going to be better than we have in CS repo.

**Comment by fagoatse** [comment]
@Swartz27 
The renderers in CoP were gimped in comparison to CS so that the game could be run on older hardware. The guy who wrote the dx11 backed(@sopyer) in actually quite active @ github so you might want to ask him for details.
Shader caching definitely improves load times but performance difference is probably negligible. Nvidia does shader caching by default on the driver level for dx10+ games since their 337.50 drivers.

**Comment by nitrocaster** [comment]
> Nvidia does shader caching by default on the driver level for dx10+ games since their 337.50 drivers.

Well, in that case, it actually becomes redundant.

**Comment by ghost** [comment]
@fagoatse : That's good to know. Even more of a reason to get rid of precompiled shaders (don't most people use SSD's at this point anyway? I imagine the difference is negligible, but I plan on comparing it when I do remove precompiled shaders).

I'd ask sopyer, and thanks for the suggestion, but wouldn't he be mad that the COP code got leaked?

@nitrocaster : I heard from K.D. as I'm having trouble with the render difference of the grass shadows implementation I ported. Anyway, in terms of working with us and also OGSE features, he had this to say: "Unfortunately, I stuck with VS2010 now. May be i should wait a little before start to work with your repo, need just fugure out, how to work with git.
When OGSE will be released there will be available source code of all features. And i'll help to work with it." So it sounds like at some point we'll have K.D. onboard, which would be epic :D

**Comment by nitrocaster** [comment]
> don't most people use SSD's at this point anyway?

Precompiled shaders make loading faster because compilation process is slow, essentially if you have tons of various shaders. There're several examples in nvidia d3d sdk that start up pretty long just because of compiling shaders.

**Comment by nitrocaster** [comment]
> I'd ask sopyer, and thanks for the suggestion, but wouldn't he be mad that the COP code got leaked?

I don't think so, he'd rather be ok with that :)

**Comment by ghost** [comment]
I'll message him then :)

Update on precompiled shaders: I successfully disabled them for the R4 renderer and tested it out. Although I did not measure loading times, they seemed to be identical to when precompiled shaders were used. However the campfire light and shadow issue was not fixed by removing precompiled shaders. I did a winmerge compare of the dx10 vs dx11 renderers, and oddly almost everything is the same, except that DX11 has a couple of tessellation references.

Very odd. Something else is causing the R1 campfire lighting issue. I'll see if sopyer knows.

Also, not strictly related to code (more to shaders) but tess_hm doesn't even work. tess_pn does of course, although the effect is weak. The height map tessellation though looks like it was not ever implemented correctly which is probably why it was never used.

**Comment by nitrocaster** [comment]
> The height map tessellation though looks like it was not ever implemented correctly which is probably why it was never used.

We'll add it to the version plan.

I have a little update: all projects are converted & fixed, and now I'm in the middle of compiling them.
LuaJIT and LuaBind have just been compiled.

This is going to be a single huge blob commit (which is considered evil), but there's no any other way.

**Comment by nitrocaster** [comment]
Done with xrAPI, xrCDB, xrCore and xrAI.

**Comment by tsf00181** [comment]
so...is CS still going to see these changes?

**Comment by nitrocaster** [comment]
What do you mean?

**Comment by fagoatse** [comment]
@tsf00181 @nitrocaster @Swartz27 
See https://github.com/nitrocaster/xray. Swartz27's fork hasn't been updated in a while. Perhaps creating a github organization instead of having everything here would be a better idea. No idea if it's free though. I suggest that you come up with a name for the project for example Openxray or Openxr or so and set up an irc chan @ freenode since there are a lot of various devs lurking there.

tsf00181 has brought up a valid point though. I haven't done any comparisons but I assume that lots of code hasn't changed/has seen minor changes in CS->CoP dev phase so whatever can be backported should be done so. A mirror of https://xp-dev.com/sc/204486/HEAD/ as a reference would be quite convenient as well.

**Comment by nitrocaster** [comment]
> ... can be backported should be done so. 

Definitely should be, you're right.

> A mirror of https://xp-dev.com/sc/204486/HEAD/ as a reference would be quite convenient as well.

I didn't get it. It's SoC repo with lots of unknown changes, not CoP.

As for github organization - good idea, I'll try to make something about.

**Comment by nitrocaster** [comment]
Ok, let's try: https://github.com/OpenXRay

**Comment by fagoatse** [comment]
@nitrocaster looks good
speaking of that soc repo, some of those changes might actually be useful for CS/CoP. From what I've seen they've added additional modding capabilites, better debug output, some nice ideas such as bloodmarks, additonal cvars. A mirror of that repo would be nice to have for those who don't want to deal with that terrible piece of crap that svn is.

**Comment by nitrocaster** [comment]
+1 for svn :)

**Comment by awdavies** [comment]
Alright, if no one has started, I'll try to register a freenode channel.  #openxr or #openxray

**Comment by ghost** [comment]
Good idea.
I'm going to leave the source up on this page, but I'm going to update it to redirect people to OpenXRay as we want to be using that (as it will be on VS2013). The only problem I have with VS2013 is KD said he can't use it, I'm assuming because he's on an older operating system. He's such a good programmer and shader author though that I'm tempted to offer to buy him a newer OS :P 

@fagoatse: I'm actually using a lot of stuff from that repo (changes had to be made as code has changed, but lots of it can be used). Grass shadows, actor shadow, etc. come from there. The only reason I haven't uploaded any of these changes is because I've been waiting on nitrocaster to update to 2013 for us. As soon as he does I'll commit the changes as long as we're in agreement about them when that time comes.
As for bloodmarks on bodies, I couldn't get that to work in CS or COP. I think what I need to look at is the SkeletonWallmark functions, and compare them to SOC to see what has changed that made them stop working.

**Comment by awdavies** [comment]
If you want to buy him an OS, I'd gladly pitch in :)

He's a swell guy.

**Comment by awdavies** [comment]
Okay, so according to freenode policy we have to have some sort of ownership over the namespace that we're going to be registering (if we want to make a primary channel).  Given that there is a github organization, I will assume for now that this is going to be okay.

We now have two channels #openxray, and #openxray-offtopic registered on freenode.

**Comment by nitrocaster** [comment]
Good, thank you.

**Comment by awdavies** [comment]
Anyone who would like to be added to the list of founders for the #openxray channel, please message me your freenode nick (or email it to me.  whichever you're comfortable with).

**Comment by andrew-boyarshin** [comment]
@awdavies, how can I email you? I doesn't see any email address on your profile page. And as far as I know, GitHub has no messaging system.

**Comment by awdavies** [comment]
I've updated my profile to have my email address.  Looks like it wasn't there for whatever reason.

Anyway, being a founder is mainly for administrative stuff like setting topics and such.  According to freenode STALKER_2010 is not registered.

**Comment by awdavies** [comment]
@fagoatse I agree.  I've wanted to migrate the SoC repo over to git, but I'm currently not sure what the best approach is, since we'd have to convince all the contributors to switch over to git as well (unless we just plan on patching in some code from the SoC repo?)

**Comment by nitrocaster** [comment]
Ok, I've got the entire solution compiled in VS2013 and tested single and non-dedicated multiplayer. The only problem now is dedicated server (crashes on start), I'm working on it.

**Comment by nitrocaster** [comment]
To anybody who compiled these sources in VS2008/VS2010:
Do you have tons of "ID3DDepthStencilState #%d created." messages in log while running debug configuration? I wonder if it's some new bug that I've introduced or it was already there.

**Comment by awdavies** [comment]
@nitrocaster Excellent!  I can't help with the debug messages, though.  When I was working with SoC, I couldn't even get it to compile with the debug configuration.

**Comment by awdavies** [comment]
I signed @nitrocaster for the time being.

**Comment by awdavies** [comment]
Assigned\* (sorry, on my phone...)

**Comment by nitrocaster** [comment]
I've pushed VS13-related huge update, so you may start to explore it.
Because 3rd party libraries were removed, you need to get them first to compile the engine. I'll add build instructions with all required explanations soon, stay updated.

**Comment by BeaconDev** [comment]
Can someone please add me to the OpenXRay page?

**Comment by ghost** [comment]
I'd like to be present in OpenXRay page too.

**Comment by Highflex** [comment]
Same here, i started digging into the game code and porting maya plugins to 2014/2015

**Comment by andrew-boyarshin** [comment]
+1. Add me please to the OpenXRay organization. Starting porting X-Ray to newer LuaJIT and LuaBind.

**Comment by andrew-boyarshin** [comment]
@nitrocaster Which codecs should we use? We need Vorbis, OGG, Theora. Where can we get them?

**Comment by nitrocaster** [comment]
@STALKER2010 as I mentioned above, I'll clarify this ASAP. Wait a little.

**Comment by andrew-boyarshin** [comment]
Aww... Can't wait... I want to compile it as fast as possible! ;) Hope, that it is easy enough... I can find codecs, but MagicSoftware stuff I can't find(required for xrAPI).

**Comment by nitrocaster** [comment]
Another thing: technically, the last huge commit I made is **bad** and we must not make such commits in future. Now I'm starting to think I could split it ... but whatever, it's really good now so we can start our work together.

**Comment by andrew-boyarshin** [comment]
Must have for us: [PVS-Studio](http://www.viva64.com/ru/pvs-studio/). I've already checked some projects, and only xrCore gives 20(!) 1-level warnings.

**Comment by nitrocaster** [comment]
Here's an [archive with all required libraries](https://dl.dropboxusercontent.com/u/63973281/dev/xray/xray-16-libraries.7z)
(temporarily on dropbox)

**Comment by nitrocaster** [comment]
Build instructions: [doc/howto/build.txt](https://github.com/OpenXRay/xray-16/blob/master/doc/howto/build.txt)

**Comment by ghost** [comment]
Bugtrap projects (minizip, zlib, bugtrap) are missing from solution?

**Comment by nitrocaster** [comment]
Yes, I forgot to include it. Have just fixed.

**Comment by ghost** [comment]
I can't compile any of the xrRender files from the latest stuff, errors are pointing to files related to actor shadow.

**Comment by nitrocaster** [comment]
Same problem. I gonna revert that commit until we get it working.

**Comment by nitrocaster** [comment]
Fixed.

**Comment by ghost** [comment]
Now i could successfully compile with release configuration, but with June 2010 DirectX SDK i get the same Dx11 shader error at game startup as before:
! D3DReflectShader hr == 0x80004002

**Comment by nitrocaster** [comment]
That's not related to compilation. Move to #openxray and we'll continue there.

**Comment by ghost** [comment]
What is  #openxray?
Well, i guess i have to wait for someone who will fix this issue with Dx SDK June 2010 and Dx11.

**Comment by nitrocaster** [comment]
#openxray is our IRC channel on freenode.org.

As for the error - I've found the reason, working on it.

**Comment by andrew-boyarshin** [comment]
@nitrocaster I have troubles with compiling. dxt.lib - what project creates it? I can't find it and can't compile solution.

**Comment by adoug** [comment]
I think you need to compile utils\DXT project before xcLC

**Comment by ghost** [comment]
"Fixed paths and names for compiling. Don't forget to correctly set path var XRAY_16X_LIBS to libs folder(to nitrocaster's zipped libs folder). If it fails yours installation's compilation proccess, revert it or optimize for it."

This one latest fix can fix that problem, that #include lines, which referring to XRAY_16X_LIBS folders headers, are not detected properly by VS2013? The libs folders are there, but VS doesn't find the included header files (especially openautomate, vorbis, ogg, theora).

The bugtrap projects are still not included in the engine.sln.

**Comment by andrew-boyarshin** [comment]
So... R3's shaders seem to be broken. When they'll be fixed? Other renders work great, everything compiled successfully.

**Comment by andrew-boyarshin** [comment]
@Bangalore1010 Hmmm... have you restarted VS after env variable edition?

**Comment by ghost** [comment]
@STALKER2010 : i restarted VS multiple times after setting the XRAY_16X_LIBS env. variable, but it was yesterday, i didn't try today's commits yet.

**Comment by nitrocaster** [comment]
> The bugtrap projects are still not included in the engine.sln.

It have not to be included in the solution since it's original 3rd party library.

> So... R3's shaders seem to be broken.

They're not broken, the problem refers to d3dcompiler headers that exist both in DXSDK and Windows SDK. I'm working on it.

For XRAY_16X_LIBS - in my case it's system environment variable, path ends with backslash. VS detects it well.

**Comment by awdavies** [comment]
@Bangalore1010 

As @nitrocaster said, XRAY_16X_LIBS needs a trailing slash in order to function properly.  The 7zip file containing the 3rd party libs should also contain the bugtrap static library.

Ex: for XRAY_16X_LIBS

```
C:\Path\To\Libs   <-- This won't work.

C:\Path\To\Libs\  <-- this should work.
```

I just have my XRAY_16X_LIBS pointing to the "3rd party" folder in <strong>src\

**Comment by ghost** [comment]
@awdavies : thanks, that was the problem, i already fixed it, i took a wrong line as an example, without backslash.
Now i have another problem, i can't start a new game, but it could be an error on my side.

**Comment by awdavies** [comment]
@Bangalore1010 

I haven't managed to create a new game either myself.  Is it running under the Release configuration or Debug?  (If you don't know, it's probably Debug).

Have you tried a clean compile?  This crash might be due to an issue with the DX11 shaders.  @nitrocaster  recently pushed a fix and you might need to recompile xrGame (I know, it's really slow :c ).

I don't have access to my home computer right now, so I can't try any of this out yet.

**Comment by nitrocaster** [comment]
Try to run new game in Debug configuration (I haven't tried to run release yet).
Make sure your working copy is up-to-date with current master branch.

**Comment by ghost** [comment]
I started the compile from scratch (deleted all bin, lib, intermediate folder), with latest and untouched files, and Release config still doesn't work. I believed first it's because of Dx11 changes, but nope, now i can't start a new game on any renderer, with the same "parameter incorrect error 87" crash.

The compile itself is flawless and almost completely done, only 4 project fails.

**Comment by nitrocaster** [comment]
Please provide bugreport (can be found in %APPDATA%\XRay Engine\reports). Use debug configuration.

> only 4 project fails.

Why? There should be no projects that cannot be compiled. Elaborate on this.

**Comment by ghost** [comment]
I get this ctd every time with Release config:

stack trace:

0023:06A20C69 xrGame.dll, CEntity::shedule_Update(), d:\src\xrgame\entity.cpp, 328
0023:06A2371D xrGame.dll, CEntityAlive::shedule_Update(), d:\src\xrgame\entity_alive.cpp, 210
0023:068A5EFA xrGame.dll, CActor::shedule_Update(), d:\src\xrgame\actor.cpp, 1222
0023:00457965 xrEngine.exe, CSheduler::Update(), d:\src\xrengine\xrsheduler.cpp, 454
0023:06A3A700 xrGame.dll, CGamePersistent::OnFrame(), d:\src\xrgame\gamepersistent.cpp, 662
0023:0040C111 xrEngine.exe, CRegistrator::Process(), d:\src\xrengine\pure.h, 86
0023:00435AAA xrEngine.exe, CRenderDevice::FrameMove(), d:\src\xrengine\device.cpp, 486
0023:00435D09 xrEngine.exe, CRenderDevice::on_idle(), d:\src\xrengine\device.cpp, 263
0023:00435C3A xrEngine.exe, CRenderDevice::message_loop(), d:\src\xrengine\device.cpp, 380
0023:00435B8E xrEngine.exe, CRenderDevice::Run(), d:\src\xrengine\device.cpp, 424
0023:0046A92D xrEngine.exe, Startup(), d:\src\xrengine\x_ray.cpp, 383
0023:00469685 xrEngine.exe, WinMain_impl(), d:\src\xrengine\x_ray.cpp, 873
0023:004691D6 xrEngine.exe, WinMain(), d:\src\xrengine\x_ray.cpp, 942
0023:00432BF3 xrEngine.exe, __tmainCRTStartup(), f:\dd\vctools\crt\crtw32\dllstuff\crtexe.c, 618
0023:7577338A kernel32.dll
0023:77D29F72 ntdll.dll
0023:77D29F45 ntdll.dll

[error][ 87] : The parameter is incorrect.

**Comment by awdavies** [comment]
The four failing projects are due to some dependencies being out of order.  I've partially fixed this issue.  I'll need to do another check build to find out more.

The CTD is one I've seen before, but only when compiling under release configuration.  I will try to run through some break points to determine the origin of the error.

In the mean time @Bangalore1010 do you know how to set breakpoints and attach VS2013 to a running process for debugging?  It's pretty straightforward and a very powerful tool.

**Comment by awdavies** [comment]
Any CTD issues should be discussed further in #11

**Comment by creWn** [comment]
As i understand the source is not compilable? Or this is mistakes on my side?

**Comment by awdavies** [comment]
Last I checked the source was compilable.  You just have to make sure you have the correct versions of Visual Studio and corresponding libraries.  I believe there's a link somewhere about it, but I don't have it handy at the moment.  I might post later if I find it.

**Comment by creWn** [comment]
So i have two problem with compilation:
1. LINK : fatal error LNK1104: cannot open file 'dxt.lib'
2. LINK : fatal error LNK1104: cannot open file 'cximage.lib'
I check and not find match libs. Are These libs must be in my 'dev' folder? Why vs request them?

**Comment by jakebriggs** [comment]
I've managed to compile the code, it produces executables, and the script provided copys the binaries to the right place in the stalker install. But, the script moans about a missing file on  the last copy:

%cp_tool% %src%\%cfg_long%_Dedicated\dedicated\xrEngine.exe %dst%\dedicated\xrEngine.exe

Also, once its copied, how do you use them?  stalker-COP.exe is not produced by the compile, and the one in the install dir will look at bin, not _bin_dbg surely

**Comment by nitrocaster** [comment]
`dedicated\xrEngine.exe` is a dedicated server executable. You have to build it separately (see `***_Dedicated` build configurations).

Regarding `Stalker-COP.exe` - it's just a launcher that can be replaced with the following cmd script:

```
start bin/xrEngine.exe
```

It's intended that `_bin_dbg` contains new binaries that you build, while `bin` conserves original binaries.

**Comment by jakebriggs** [comment]
Okay, so the dedicated server is not needed to run the game single player. Excellent. Also, thanks for the advice on how to start the game - much appreciated!

**Comment by jakebriggs** [comment]
EDIT: setting $app_data_root$         = true | false | C:\Games\S.T.A.L.K.E.R. - Call of Pripyat\
in fsgame.ltx solved the below problem

edit2: no it didn't....

I am using visual studio 2013 community edition to compile, and I am compiling successfully, but the game won't start using start _bin_dbg/xrEngine.exe - start bin/xrEngine.exe works fine (I didn't compile that one though).

Starting it from visual studio gives me:

```
Unhandled exception at 0x61331752 (msvcr120.dll) in xrEngine.exe: 0xC00000FD: Stack overflow (parameters: 0x00000000, 0x00092000). 
```

and [this log](http://pastebin.com/nfZ1WbPw)

**Comment by nitrocaster** [comment]
Do you really get stack overflow? Could you post stack trace?

`FS: 5768 files cached 0 archives`
No one archive found. Do you have original game resources?

**Comment by jakebriggs** [comment]
Run from _bin_dbg: http://pastebin.com/UTXdVEEq

**Comment by jakebriggs** [comment]
and running from visual studio: http://pastebin.com/F6xSaLbm

**Comment by nitrocaster** [comment]
You should have got message box when the first failure happened. See `src\xrCore\xrDebugNew.cpp:220`
`nodefault reached` means that MessageBox returned some unexpected value.

**Comment by jakebriggs** [comment]
and running from visual studio:
http://pastebin.com/xcCGy1SM

**Comment by nitrocaster** [comment]
Please use pastebin.com or another external storage for these huge logs.
What's about message box I asked above?

**Comment by jakebriggs** [comment]
Sorry about that. I've set the game to be windowed. Upon starting the game, I get the little splash screen, then the window pops up - all black, then I get "X-ray 1.6 has stopped working" message:

http://i.imgur.com/aLrTHiE.png

I don't think that is the pop up you were expecting....

**Comment by jakebriggs** [comment]
I solved it. I still had stalker complete mod installed. I was sure I removed it, bit obviously I was mistaken.

Sorry for wasting your time, I feel like an idiot!

**Comment by ViHtarb** [comment]
Plz help me

I build sources from this repo and noticed that building xrGame project is a very very slow compared with build a clear sources of xrGame in Visual Studio 2008.

Example, i comment two same lines in Actor.cpp in a both projects(from this repo and in clear project) As a result of building the project time was 4 minutes(VS2013) and 2 minutes(VS2008)

How to fix it?

**Comment by nitrocaster** [comment]
Try to disable **Multiprocessor compilation** (`/MP`) and enable **Minimal rebuild** (`/Gm`).

**Comment by ViHtarb** [comment]
this not helped :(

**Comment by nitrocaster** [comment]
No way. Commenting two lines in any single cpp file means that only one cpp file should be recompiled. Given that you have precompiled headers enabled, such compilation should be quite fast.

**Comment by sobkas** [comment]
I would like to have a prebuild version of engine, I'm trying to debug wine nine state tracker code and this would help a lot

**Comment by andrew-boyarshin** [comment]
@nitrocaster @Armada651 close this one?

**Comment by CrossVR** [comment]
@andrew-boyarshin If you think an issue can be closed, go ahead and close it. If it really still is an issue someone will reopen it.
